### ISSUE TITLE
setup_environment (modules) needs more complexity

### ISSUE THREAD

**Issue by olesenm** [spack/spack#6501]
This is a partial followup to #4942, but seems to be more general deficiency.

I now have my `setup_environment()` using `EnvironmentModifications.from_sourcing_file` with a guard to ensure that this part is only sourced post-install. So far, so good (no crashes). However by doing this, there are a number of user-related settings that have essentially leaked into the modules.

In an ideal implementation, I would have this type of content:

    setenv      FOAM_USER_APPBIN "$env(HOME)/OpenFOAM/$env(USER)-$env(WM_PROJECT_VERSION)/platforms/$env(WM_OPTIONS)/bin" 

    append-path PATH "$env(FOAM_USER_APPBIN)" 

Currently, since it works directly with the environment values, these are expanded prior to building the modifications. This means that the module content includes the user-local bin directory of the person who created the module.

Now that I start to understand what is going on in the spack.environment package, I could start replicating some of its behaviour, but it doesn't look like there is a way to defer expansions. That is to say, to place information like this in the module file:

    setenv      FOAM_USER_APPBIN "$env(HOME)/OpenFOAM/$env(USER)-$env(WM_PROJECT_VERSION)/platforms/$env(WM_OPTIONS)/bin"

**Comment by olesenm** [comment]
@alalazo 

I've started making a few changes to get things working better. I have them as WIP here: 
https://github.com/olesenm/spack/commits/wip-modules

I think the additional filtering makes sense. Can provide a blacklist, whitelist of environment variables where modifications should be ignored or retained. It is an optional parameter, so I don't expect any odd side effects.

The next part deals with what happens afterwards. In sourcing the OpenFOAM bashrc, it does some extra work to clean empty and duplicate entries from the PATH. This causes the rest of the module building to gag since it looks like much more of the PATH has changed than really has. I've added in a fix for that too, but tagged it as a "WIP" since it may need some discussion.
Besides which, I think there is a bug in the handling of prepend-path. It seems that the original order of appearance it retained, but this means that order gets reversed.

For example,

    LD_LIBRARY_PATH=dir1:dir2:dir3

becomes

    prepend-path LD_LIBRARY_PATH=dir1
    prepend-path LD_LIBRARY_PATH=dir2
    prepend-path LD_LIBRARY_PATH=dir3

which results in

    LD_LIBRARY_PATH=dir3:dir2:dir1

I would like to get all these fixed in a single merge.
